arch/arm64/core/mmu.c

```diff
@@ -806,6 +806,79 @@ static uint64_t get_tcr(int el)
 	return tcr;
 }
 
+#ifdef CONFIG_DT_HAS_XEN_XEN_ENABLED
+
+static int flush_inv_mapped_pages(int (*cmo_fn)(void *, size_t),
+				  uint64_t *pte, int level, uint64_t base)
+{
+	int ret = 0;
+	uint64_t level_size = BIT(LEVEL_TO_VA_SIZE_SHIFT(level));
+	const uint64_t max_va_addr = (uint64_t)CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE;
+
+	if (!xlat_use_count[table_index(pte)]) {
+		return 0;
+	}
+
+	for (unsigned int i = 0; i < Ln_XLAT_NUM_ENTRIES; i++) {
+		uint64_t end, va = base + i * level_size;
+		uint8_t mem_type;
+
+		if (is_free_desc(pte[i])) {
+			continue;
+		}
+
+		if (is_table_desc(pte[i], level)) {
+			/* Move to the next translation table level */
+			ret = flush_inv_mapped_pages(cmo_fn, pte_desc_table(pte[i]), level + 1, va);
+			if (ret) {
+				break;
+			}
+			continue;
+		}
+
+		/*
+		 * From this point, this must be a leaf.
+		 * Start excluding non memory mappings
+		 */
+		mem_type = (pte[i] >> 2) & MT_TYPE_MASK;
+		if ((mem_type != MT_NORMAL) && (mem_type != MT_NORMAL_NC)) {
+			continue;
+		}
+
+		end = va + level_size - 1;
+
+		/* No intersection with RAM? */
+		if (end < CONFIG_KERNEL_VM_BASE || va >= max_va_addr) {
```

Comment:
How this can happen?

Reply:
Well, U-Boot has own quirks. 

This is why you should not mindlessly copy code from one project to another.

Reply:
Who knows maybe here we can have the same problems

Reply:
So, let me sum this up. You authored this patch by taking code from another project, with incompatible license and you don't fully understand what this code does. But you wish to include this code into Zephyr mainline. Is this correct?

Reply:
I understand what this code does, but I 'm not sure that Zephyr doesn't have exact the same problems as in u-boot
maybe somebody writes some bad address to table or maybe memory with page table become corrupted

Reply:
Or maybe they will do the correct thing, but you code will miss this, because of the extra unneeded check?

Look. This is the MMU driver. It is one of the most crucial things in any OS. You need to take extra care when modifying it. And you need clear understanding of what it does, how it works and what are consequences of your changes, Anyone, who will read your patch will ask self "why does @xakep-amatop  added this check? what is the intention? Why there are no similar checks in the other parts of the driver?" For example, `set_mapping` also traverses pagetables in the similar fashion, but it has no similar checks.

It is normal to guard against programer errors. But there already is a mechanism for this: `ASSERT`.  It clearly states your intention to guard against mistakes. Take a look at `z_mem_phys_addr` for example. Your code, in turn, does not prevent mistakes. It just silently ignores them. Is it right thing to just `continue` if there is a bad address or pagetable corruption?

Reply:
> why does @xakep-amatop added this check?

Reply:
who knows it was tested well enough or not maybe there is here some programmers errors
as far as I see, arm64 isn't so popular architecture now as some others and how many people tested it with enabled MMU or not

Reply:
set panic instead of continue and caught crash